### PR TITLE
Input validation matrix

### PR DESCRIPTION
**Pull request by dktcoding**
I used exceptions for the constructors and asserts for most of the input validation as discussed, except for `Matrix#dotProduct()` where the inputs could match in dimensions, but be incorrect for the dot product itself.

### CODE REVIEW

File: src/libai/common/Matrix.java [hunk 1]
```diff
@@ -248,67 +306,70 @@ public void applyInIdentity(Function f, final Matrix a) {
...
+	 * {@code is false}, and (-1, 1) if {@code true}.
 	 *
-	 * @param signed {@code false} if all the numbers should be positive, 
+	 * @param signed {@code false} if all the numbers should be positive,
 	 * {@code false} otherwise
-	 * @see Matrix#fill(boolean, java.util.Random) 
+	 * @see Matrix#fill(boolean, java.util.Random)
 	 */
 	public void fill(boolean signed) {
         fill(signed, null);
```
Comment (kronenthaler) [inline src/libai/common/Matrix.java]: indentation seems off

File: src/libai/common/Matrix.java [hunk 2]
```diff
@@ -320,8 +381,24 @@ public void copy(final Matrix a) {
 	 *
 	 * @param a The matrix to multiply
 	 * @return the scalar of the dot product.
+	 * @throws IllegalArgumentException if either {@code this} or {@code a} are
+	 * neither row nor column matrices.
 	 */
 	public double dotProduct(Matrix a) {
+		assert a != null : "a must be not null";
+		assert rows * cols == a.rows * a.cols :
+			   "a doesn't have the right dimensions";
```
Comment (kronenthaler) [inline src/libai/common/Matrix.java]: mismatched dimensions sounds more professional

File: src/libai/common/Matrix.java [hunk 3]
```diff
@@ -332,15 +409,16 @@ public double dotProduct(Matrix a) {
...
+	 * <p>The matrix a must be created and has the right dimensions.</p>
+	 * <p><i>NOTE:</i> Assertions of the dimensions are made with {@code assert}
+	 * statement. You must enable this on runtime to be effective.</p>
 	 *
 	 * @param a The matrix to put the result.
 	 */
 	public void transpose(final Matrix a) {
-		assert rows == a.cols && cols == a.rows;
+		assert rows == a.cols && cols == a.rows :
+			   "Mattrix a has the wrong dimensions";
```
Comment (kronenthaler) [inline src/libai/common/Matrix.java]: typo, double t in matrix

File: src/libai/common/Matrix.java [hunk 4]
```diff
@@ -384,66 +466,87 @@ public void setValue(double v) {
...
+	 * statement. You must enable this on runtime to be effective.</p>
 	 *
 	 * @param index The index of the column to return.
 	 * @return the array with the values.
 	 */
 	public double[] getCol(int index) {
-		double[] ret = new double[rows];
-		for (int i = 0; i < rows; ret[i] = position(i++, index));
+		assert index < 0 || index >= cols : String.format(
+			  "j must be in the interval [0, %d) current: %d", cols, index);
```
Comment (kronenthaler) [inline src/libai/common/Matrix.java]: there is no j variable in this method

File: src/libai/common/Matrix.java (hunk 1 above)
Comment (dktcoding) [inline src/libai/common/Matrix.java]: Oh, darn soft tabbing, always forget to disable it

File: src/libai/common/Matrix.java [hunk 5]
```diff
@@ -384,66 +465,87 @@ public void setValue(double v) {
...
+	 * statement. You must enable this on runtime to be
+	 * effective.</p>
 	 *
 	 * @param index The index of the row to return.
 	 * @return the array with the values.
 	 */
 	public double[] getRow(int index) {
-		double[] ret = new double[cols];
+		assert index < 0 || index >= rows : String.format(
+			  "i must be in the interval [0, %d) current: %d", rows, index);
```
Comment (kronenthaler) [inline src/libai/common/Matrix.java]: index must be in the interval

File: src/libai/common/Matrix.java [hunk 6]
```diff
@@ -559,15 +681,15 @@ public void printOOFormat(PrintStream out) {
 	@Override
 	public boolean equals(Object b1) {
 		if (this == b1) {
-		    return true;
+			return true;
 		}
 		if (b1 == null) {
```
Comment (kronenthaler) [inline src/libai/common/Matrix.java]: could we compress these 2 conditions into one single if statement

```
if (b == null || getClass() != b1.getClass()){ 
    return false; 
}
```

File: src/libai/common/Matrix.java [hunk 7]
```diff
@@ -559,15 +681,15 @@ public void printOOFormat(PrintStream out) {
...
-		    return false;
+			return false;
 		}
 		if (getClass() != b1.getClass()) {
-		    return false;
+			return false;
 		}
-        
+
 		Matrix b = (Matrix) b1;
```
Comment (kronenthaler) [inline src/libai/common/Matrix.java]: maybe include this statement too in the if

File: src/libai/common/Matrix.java (hunk 6 above)
Comment (dktcoding) [inline src/libai/common/Matrix.java]: Yeahp, it will look much cleaner, I'll change it when I get home

**Pull request merged**